basement/Basement/String.hs

```diff
@@ -1321,56 +1322,62 @@ decimalDigitsPtr startAcc ptr !endOfs !startOfs = loop startAcc startOfs
-                             else return Nothing
+                              then Just ns
+                              else Nothing
             | otherwise = do
-                let !(c, idx') = nextI idx
-                    !cm@(CM c1 c2 c3) = op c 
+                let !(Step c idx') = UTF8.next ba idx
+                    !(CM c1 c2 c3) = op c 
                     !cSize = if c2 == '\0' -- if c2 is empty, c3 will be empty as well.
-                              then charToBytes (fromEnum c1) 
-                              else eSize c1 + eSize c2 + eSize c3
+                                then charToBytes (fromEnum c1) 
+                                else eSize c1 + eSize c2 + eSize c3
                     !nchanged = changed || c1 /= c || c2 /= '\0'
                 loop idx' (ns + cSize) nchanged
 
 -- | Convert a 'String' 'Char' by 'Char' using a case mapping function. 
 caseConvert :: (Char -> CM) -> String -> String
-caseConvert op s@(String ba) 
-  = case nBuff of
+caseConvert op s@(String arr) = case nBuff of
       Nothing -> s
-      Just nLen -> runST $ unsafeCopyFrom s nLen go
+      Just nLen -> runST $ do
+          ((), dst) <- newNative nLen $ \mba ->
+                  C.onBackendPrim
+                        (\blk  -> goBlk mba blk (Offset 0) start)
+                        (\fptr -> withFinalPtr fptr $ \ptr -> goBlk mba ptr (Offset 0) start)
+                        arr
+          freeze dst
   where
+    !(C.ValidRange start end) = C.offsetsValidRange arr
     !nBuff = caseConvertNBuff op s
-    go :: String -> Offset Char -> Offset8 -> MutableString s -> Offset8 -> ST s (Offset8, Offset8)
-    go src' srcI srcIdx dst dstIdx = do
-      let !(CM c1 c2 c3) = op c 
-      dstIdx <- write dst dstIdx c1
-      nextDstIdx <- 
-        if c2 == '\0' -- We don't want to check C3 if C2 is empty.
-          then return dstIdx
-          else do
-            dstIdx  <- writeValidChar c2 dstIdx
-            writeValidChar c3 dstIdx
-      return (nextSrcIdx, nextDstIdx)
-          where
-            !(Step c nextSrcIdx) = next src' srcIdx
-            writeValidChar cc wIdx =
-                if cc == '\0'
-                    then return wIdx 
-                else do
-                    write dst wIdx cc
+    goBlk :: (Indexable container Word8, PrimMonad prim) => MutableBlock Word8 (PrimState prim) -> container -> Offset Word8 -> Offset Word8 -> prim ()
+    goBlk !dst !src = loop
+      where
+        loop !dstIdx !srcIdx
+          | srcIdx == end = return ()
+          | otherwise = do
+              let !(CM c1 c2 c3) = op c 
+                  !(Step c nextSrcIdx) = UTF8.next src srcIdx
+              !dstIdx <- UTF8.writeUTF8 dst dstIdx c1
+              nextDstIdx <- 
+                if c2 == '\0' -- We don't want to check C3 if C2 is empty.
+                  then return dstIdx
+                  else do
+                    dstIdx  <- writeValidChar c2 dstIdx
```

Comment:
Can't we use `UTF8.writeUTF8 dst dstIdx c2` directly instead of `writeValidChar c2 dstIdx`?
At this point, we already checked `c2 == '\0'` and the call to writeValidChar would check that again.

Reply:
as a disclaimer, I barely look at the code when rebaseing it to master. it look likely that's it redundant.

Reply:
You're absolutely right.